rancher2/resource_rancher2_cluster.go

```diff
@@ -537,6 +537,7 @@ func findFlattenClusterRegistrationToken(client *managementClient.Client, cluste
 }
 
 func findClusterRegistrationToken(client *managementClient.Client, clusterID string) (*managementClient.ClusterRegistrationToken, error) {
+	log.Printf("[TRACE] Finding cluster registration token for %s", clusterID)
 	for i := range clusterRegistrationTokenNames {
 		regTokenID := clusterID + ":" + clusterRegistrationTokenNames[i]
 		for retry, retries := 1, 10; retry <= retries; retry++ {
```

Comment:
Could we try 3-4 times to find the cluster registration token before we use an existing? Did retrying 10 times work better in testing?

Reply:
We try 10 times for each token (2 max, generally 1 though) before creating a new one. This works better IMO, as we want to use the generated one, but occasionally, trying to create it would fail because the generated one would be created before we could create one. I think in that case it's better to just return that one, so the total logic is as follows:
- for each token
  - try ten times to get the token
    - if we found it, return it
    - if it doesn't exist, sleep 3 seconds before trying again
    - if any other error, return
- if no tokens were found
  - create a new token, and return it
  - if this token fails to create because it already exists, return that
  - any other error, return
We might even do well to bump the retries up to 15 to prevent thrashing, but I'm not sure of the potential side effects of not waiting long enough to create the token.

Reply:
Ok, that sounds fine to me then

Reply:
Yes, that correct. The other issue is that these operations can't be done atomically, i.e. we can wait forever, but we still run the risk of rancher generating the token after we try to create it, which is why the change was made to just return that in the event it already exists.